### PR TITLE
BE: Messaging

### PR DESCRIPTION
**Pull request by HAZberi**
### What this PR does (required):
- Create database model, controllers and routes for chat messaging. 

### Screenshots / Videos (required):

https://user-images.githubusercontent.com/72773994/126213560-3b159b04-60df-4ad4-8685-6258ad34c2cb.mov
### Any information needed to test this feature (required):
- Chat routes can be tested with Postman

### Any issues with the current functionality (optional):
- N/A

### CODE REVIEW

File: server/models/Message.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+    senderID: {
+        type: mongoose.Schema.Types.ObjectId,
+        ref: 'user',
+        required: true
+    },
+    message: {
+        type: String,
+        required: true,
+    },
+}, {timestamps: true});
```
Comment (rajivtitus) [inline server/models/Message.js:21]: 👍🏽

File: server/models/Message.js [hunk 2]
```diff
@@ -0,0 +1,20 @@
+const mongoose = require("mongoose");
+
+const messageSchema = new mongoose.Schema({
+    conversationID: {
```
Comment (rajivtitus) [inline server/models/Message.js:4]: Just a suggestion, maybe you can have `read` status as well?

File: server/controllers/message.js [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    const { senderID, convoID, message } = req.body;
+    const newMessage = new Message({
+        conversationID: convoID,
+        senderID,
+        message,
+    });
+
+    try {
+        const savedMessage = await newMessage.save();
+        res.status(200).json(savedMessage);
```
Comment (rajivtitus) [inline server/controllers/message.js]: Please use appropriate status code for resource creation. 

Ref: https://www.restapitutorial.com/httpstatuscodes.html

File: server/controllers/conversation.js [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+        res.status(200).json(savedConvo);
+    } catch (error) {
+        res.status(500).json(error);
+    }
+})
+
+//Based on userID in the params, get all conversations of the user
+
+exports.getConversations = asyncHandler(async (req, res) => {
+    const { userID } = req.params;
```
Comment (rajivtitus) [inline server/controllers/conversation.js]: Please protect your routes using the `auth` middleware. 
Once done, you can access the user's id in the request object.

File: server/controllers/conversation.js [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+// Given two userIDs, create a conversation
+exports.createConversation = asyncHandler(async (req, res) => {
+    const { senderID, receiverID } = req.body;
+    const newConvo = new Conversation({
+        participants: [senderID, receiverID],
+    });
+
+    try {
+        const savedConvo = await newConvo.save();
+        res.status(200).json(savedConvo);
```
Comment (rajivtitus) [inline server/controllers/conversation.js]: Please see below for resource on status codes.

File: server/controllers/conversation.js [hunk 6]
```diff
@@ -0,0 +1,33 @@
+const Conversation = require("../models/Conversation");
+const asyncHandler = require("express-async-handler");
+
+
+// Given two userIDs, create a conversation
+exports.createConversation = asyncHandler(async (req, res) => {
+    const { senderID, receiverID } = req.body;
+    const newConvo = new Conversation({
```
Comment (rajivtitus) [inline server/controllers/conversation.js:18]: With the current logic, multiple conversations can be created between the same two users. I don't think that's what we need? :)

**Review by rajivtitus — COMMENTED** [review]
Looks good for a start! 

I've made some comments, please have a look. Thank you!

**Comment by HAZberi** [comment]
Please review the changes as suggested.

File: server/controllers/conversation.js [hunk 7]
```diff
@@ -2,25 +2,33 @@ const Conversation = require("../models/Conversation");
...
 
 
-// Given two userIDs, create a conversation
+// Given two unique userIDs, create a private conversation
 exports.createConversation = asyncHandler(async (req, res) => {
-    const { senderID, receiverID } = req.body;
+    const senderID = req.user.id;
+    const { receiverID } = req.body;
+
+    if(senderID === receiverID) {
```
Comment (rajivtitus) [inline server/controllers/conversation.js]: We need to check whether a conversation between the two users already exists. Only if it doesn't exist, we need to create a new conversation

**Review by rajivtitus — COMMENTED** [review]
Thanks for making requested changes.

Just one comment, please review

File: server/controllers/conversation.js (hunk 7 above)
Comment (rajivtitus) [reply server/controllers/conversation.js]: You can remove this if/else construct as we are going to be sending unique sender and receiver id's from the FE

**Review by rajivtitus — APPROVED** [review]
LGTM! Thanks for making requested changes

Please address the small comment before merging

File: server/db.js [hunk 8]
```diff
@@ -1,6 +1,6 @@
 const mongoose = require("mongoose");
 
-require('dotenv').config();
+//require('dotenv').config();
```
Comment (rajivtitus) [inline server/db.js]: I just noticed this, please uncomment/remove before merging :)

**Comment by HAZberi** [comment]
Removed unnecessary if/else construct and comments. Resolved two merge conflicts as well.

**Pull request merged**